feat: show suggested registries in Podman Desktop

cdrage opened the pull request:
feat: show suggested registries in Podman Desktop

### What does this PR do?

* Shows suggested registries when going to add a registry
* Adds an API call to "suggest" a registry which can be done by any
  extension added (for example, a custom extension that provides a
  private registry to be added)
* Show base64 encoded logo (no need to use fs / async / await load a
  logo, since they are small 40x40px icons
* UI/UX rework of the registry page to make sure that login works
  correctly when adding a new registry

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

![Screenshot 2023-01-19 at 3 41 47 PM](https://user-images.githubusercontent.com/6422176/213555893-2b8210cb-ba88-40af-84d9-fa42125f9d01.png)


https://user-images.githubusercontent.com/6422176/213349477-979ee76f-1468-4663-8e49-c21cd6b276cf.mov



### What issues does this PR fix or reference?

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

Closes https://github.com/containers/podman-desktop/issues/895

### How to test this PR?

<!-- Please explain steps to reproduce -->

`yarn watch`

Go to registry and try it out! Login to Docker Hub / Quay / etc.

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

benoitf reviewed (commented):
Icon is the CRC icon so we should change it for this extension ( or drop it)

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,66 @@
+import * as extensionApi from '@tmpwip/extension-api';
+
+export async function activate(extensionContext: extensionApi.ExtensionContext): Promise<void> {
+  // Create the "provider"
```
IMHO we don't need to register a provider

cdrage replied:
Got it! I'll remove that.

benoitf replied:
👍

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,66 @@
+  {
+    name: 'Docker Hub',
+    url: 'docker.io',
+    icon: 'iVBORw0KGgoAAAANSUhEUgAAAB4AAAAeCAYAAAA7MK6iAAABhmlDQ1BJQ0MgcHJvZmlsZQAAKJF9kT1Iw1AUhU9TpSIVwXYQcchQnSyIioguWoUiVAi1QqsOJi/9gyYNSYqLo+BacPBnserg4qyrg6sgCP6AuLo4KbpIifclhRaxXni8j/PuObx3HyDUSkyzOkYBTbfNZDwmpjOrYuAVPvQhhGnMyMwy5iQpgbb1dU/dVHdRntW+78/qUbMWA3wi8SwzTJt4g3hy0zY47xOHWUFWic+JR0y6IPEj1xWP3zjnXRZ4ZthMJeeJw8RivoWVFmYFUyOeII6omk75QtpjlfMWZ61UYY178hcGs/rKMtdpDSKORSxBgggFFRRRgo0o7Top ...
```
We should read from files stored locally as it's easier to maintain

cdrage replied:
Yeah, I was debating about doing it via icon.png vs base64, since the icons were so small. I'll load it locally instead!

cdrage commented:
> Icon is the CRC icon so we should change it for this extension ( or drop it)

dropping it unfortunately results in a bug / error of podman desktop trying to load a missing `icon.png` so I'll most likely replace it with a blank one / fix the bug.

benoitf commented:
yes probably we should provide blank one if not defined

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,74 @@
+  {
+    name: 'Docker Hub',
+    url: 'docker.io',
+    icon: base64_encode(imagePath + 'docker.png'),
```
```suggestion
    icon: base64_encode(path.resolve(imagePath, 'docker.png')),
```

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,74 @@
+
+// Function that takes in a file location and returns a base64 encoded string of the file
+function base64_encode(file: string): string {
+  const bitmap = fs.readFileSync(file);
```
you may switch to asynchronous `fs.promises.readFile`

cdrage commented:
Updated!

* Resolved suggestions
* Added ability to reactively remove / add suggested registries when enabling and disabling the extension
* feat: When adding a new registry, we will check the "suggestedRegistries" list and add the icon and appropriate name instead of the URL

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,74 @@
+    icon: base64_encode(imagePath + 'github.png'),
+  },
+  {
+    name: 'IBM Container Registry',
+    url: 'icr.io',
+    icon: base64_encode(imagePath + 'ibm.png'),
+  },
```
I'm not sure about this one by default ?

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,74 @@
+    name: 'GitHub',
+    url: 'ghcr.io',
+    icon: base64_encode(imagePath + 'github.png'),
+  },
```
if github probably needs the one from Gitlab registry.gitlab.com

benoitf commented on extensions/registries/src/extension.ts:
```diff
@@ -0,0 +1,74 @@
+    icon: base64_encode(imagePath + 'ibm.png'),
+  },
+  {
+    name: 'Google Container Registry',
+    url: 'gcr.io',
+    icon: base64_encode(imagePath + 'gcr.png'),
+  },
```
here I don't know as gcr is replaced by artifact registry and if there is Google registry, probably we should have amazon and azure one as well

cdrage replied:
yeah, maybe remove IBM suggestion, add amazon and azure?

benoitf replied:
I think that we may need some enhancement on the suggesting API as for some, we need extra information

for example with Azure URLs are like `MY-CUSTOM-NAME.azurecr.io` (URL azurecr.io is not used directly)

so maybe we start with very few like GitHub, DockerHub and Quay.io and wait feedback
If people have other thoughts as well here.

cdrage replied:
Same with GitLab, so can't add it just yet (the docs show: registry.gitlab.example.com) https://docs.gitlab.com/ee/user/packages/container_registry/

You're right. Let's just do Quay, Google, Github and Docker for now.

benoitf commented on packages/main/src/plugin/image-registry.ts:
```diff
@@ -146,6 +147,33 @@ export class ImageRegistry {
+    // this may have been done by another extension.
+    if (this.suggestedRegistries.find(reg => reg.url === registry.url && reg.name === registry.name)) {
+      // Ignore and don't register
+      console.log('Registry already registered: ' + registry.url);
```
```suggestion
      console.log(`Registry already registered: ${registry.url}`);
```

benoitf commented on extensions/registries/package.json:
```diff
@@ -0,0 +1,36 @@
+  "displayName": "Registries",
+  "description": "Package a list of default registries for Podman Desktop to consume such as Quay, DockerHub, GitHub, etc. All of which will be listed under the Registries section of Podman Desktop settings.",
+  "version": "0.0.1",
+  "icon": "icon.png",
```
should be removed if there is no 'icon.png' file

benoitf commented:
I'm not sure if we need the sentence `Registry suggested by the extension` ?

cdrage commented:
Updated based on review!
* Removed IBM (due to naming issues with jp.icr.io, us.icr.io, etc.)
* "Registry suggested by the extension" removed
* Loading files is now asynchronous

cdrage commented:
```
export function suggestRegistry(registry: RegistrySuggestedProvider): void;
```

Switching to Disposable will come next commit. Trying to figure out why it's not "pushing" it correctly.

cdrage commented:
Updated with disposable instead of void!

Should I still implement the `unsuggestRegistry` function @benoitf ? Or Would the "Disposable" function cover that functionality instead? Unsure if both should be implemented into the API 👍

benoitf commented:
@cdrage you can remove the unsuggest exposed in the API so it's only managed by disposable

benoitf reviewed (changes requested):
it works well in development mode
![image](https://user-images.githubusercontent.com/436777/213991897-f202f2d9-5f44-4918-bc6d-0c3bae965419.png)

There are few nits that could be done also (removal of unused dependency / unused command)

but I tried a 'production build' and it doesn't load the extension

I think it's because the `build:extensions` target of `package.json` is not containing the build of this extension so this one is more problematic

slemeur commented:
Should we display "Registry suggested by extension" ? I'd expect the user may not know what "extension" is in this context.
"Suggested registry" would be enough for me.

slemeur commented:
Agree @cdrage, the repeat on multiple lines feels weird.
Could we remove the sentence and replace the "Add" button by "Configure" ? wdyt @mairin as an interim solution?

mairin commented:
> Agree @cdrage, the repeat on multiple lines feels weird. Could we remove the sentence and replace the "Add" button by "Configure" ? wdyt @mairin as an interim solution?

Interim, sure. I would suggest instead of having "Suggested registry" repeated in the username column (it's not a username so it could confuse for a second), we just put a "_(Suggested_)" string after the name of the registry in the Registry Location column? 

For the longer-term approach:

The original intent with this design was to only show configured registries in the table, and if you wanted an additional one, you'd press the "add registry" button in the lower right. We should have a limited number of suggested-but-not-configured registries in the table, so I'd think the impact of having repeated Add buttons would be minimal - we'd only want those registries we really want to highlight to users as being available here.

If this set of 4 is those we really want to highlight:

There could be alternative ways to suggest a registry outside of the table. For example, we could do a dismissable block maybe under the table or above it (e.g. like the Patternfly alert pattern shown here: https://www.sketch.com/s/729c2eee-e8b6-4fcd-8a79-f6faa8c30f89/a/1K3WJWk#Inspect) that could be use to suggest specific registries. Something like (for an alert):

(i) We recommend adding the following registries:
   - [icon] Docker (_add now_)
   - [icon] Red Hat Quay.io (_add now_)
   - [icon] GitHub (_add now_)
   - [icon] Google Container Registry (_add now_)
   
 Could also do a suggested registries table or block below that has a dismiss button on it? 
   
If they're built into the main table, there should be a way to dismiss ("I don't use GitHub registry, I don't have plans to, keep it out of my table" sory of use case. I think the approach should be an onboarding type one where we suggest them the first few times you visit the screen and you can dismiss it after you settle into using the app.

Does that make sense?

cdrage commented:
Updated @mairin @slemeur @benoitf with the following (until we get a final design?):

![Screenshot 2023-01-23 at 1 08 15 PM](https://user-images.githubusercontent.com/6422176/214116507-2cc236ea-3c93-4e06-94a5-508176ac4369.png)

deboer-tim commented:
+1 to Mairin's comments above. I think this would work much better not as part of the table, but separate actions underneath that you can act on or dismiss/collapse. e.g. a "Suggested repositories" section that has tiles with larger icon and name for each.

mairin commented:
Here's another idea on how to show it (although I think there should be logic perhaps to make sure we don't suggest a registry that's already there? so the example here is bad in that ideally quay and dockerhub shouldn't be advertised to this user; they already have it configured.)

![Registry Tab Quay Edit-11](https://user-images.githubusercontent.com/799683/214148263-4ff3a7eb-8a95-4f21-9519-99dee2c89aa5.png)

cdrage commented:
That's awesome @mairin ! And luckily the logic is already there in the code. I was hoping that we'd get the backend code / logic in and then we could start iterating on the design process in a separate PR. The bottom section in that mock-up will be easily doable. 

I think we should move forward with this PR @benoitf @deboer-tim as it's already quite large (over 500 lines), I'll open up another issue to update the design.

benoitf commented:
I would remove for now the `(Suggested)` or fix the height of rows

As it's displayed with 2 lines (too long), clicking on configure is moving the lines when clicking on `Configure`

https://user-images.githubusercontent.com/436777/214237025-f75fd299-ed0f-4f3c-9eee-6028d79da3da.mp4

deboer-tim commented:
I was figuring Charlie would request that once I saw the size of the PR ;-). +1 to Florent's comments or anything else minor to reduce any ugliness or disable the new UI just before merging, then making sure we follow through with improved UI soon.

cdrage commented:
> I would remove for now the `(Suggested)` or fix the height of rows
> 
> As it's displayed with 2 lines (too long), clicking on configure is moving the lines when clicking on `Configure`
> 
>  HiGeH28Wcu.mp4

I've gone ahead and remove (suggested) as there's no point fixing the height of the rows since we're going to be implementing the newer suggested / tabbed section below.

benoitf commented:
there is still a move of the line when clicking on the configure button 

https://user-images.githubusercontent.com/436777/214320577-f85c495b-0925-492c-903b-15b009e16c26.mp4

cdrage commented:
@benoitf Fixed! Took a while, but found the div that wasn't added over. It no longer changes when pressing configure / add.

https://user-images.githubusercontent.com/6422176/214332937-eb1d9050-0d32-4bc4-8393-49b4583b1390.mov

benoitf commented:
let's merge then (after fixing the PR check ;-)
